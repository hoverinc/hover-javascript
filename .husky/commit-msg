#!/bin/sh
. "$(dirname "$0")/_/husky.sh"

node src commit-msg
