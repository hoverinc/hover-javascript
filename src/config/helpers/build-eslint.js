const {
  rules: airbnbRules,
} = require('eslint-config-airbnb-typescript/lib/shared')

const {hasAnyDep} = require('../../utils')
const {testMatch} = require('../jest.config')

const withBaseConfig = base => variant =>
  require.resolve(base + (variant ? `/${variant}` : ''))

const airbnb = withBaseConfig('eslint-config-airbnb-typescript')
const prettier = withBaseConfig('eslint-config-prettier')

const hasReact = hasAnyDep('react')

const parserRules = (typescript = false, react = false) => {
  const isOff = off => (off ? 'off' : 'error')

  const propTypes = react ? {'react/prop-types': isOff(typescript)} : {}

  return {
    'no-implied-eval': isOff(typescript),
    'no-throw-literal': isOff(typescript),
    '@typescript-eslint/no-implied-eval': isOff(!typescript),
    '@typescript-eslint/no-throw-literal': isOff(!typescript),
    ...propTypes,
  }
}

const buildConfig = ({withReact = false} = {}) => {
  const isReact = withReact || hasReact
  const ifReact = (t, f) => (isReact ? t : f)

  return {
    plugins: ['prettier', 'jest', ifReact('react-hooks')].filter(Boolean),
    extends: [
      ifReact(airbnb(), airbnb('base')),
      'plugin:jest/recommended',
      prettier(),
      prettier('@typescript-eslint'),
      ifReact(prettier('react')),
      ifReact('plugin:react-hooks/recommended'),
    ].filter(Boolean),
    rules: {
      'import/prefer-default-export': 'off',
      'import/no-extraneous-dependencies': [
        'error',
        {
          devDependencies: airbnbRules[
            'import/no-extraneous-dependencies'
          ][1].devDependencies.concat([
            'jest/**',
            'e2e/**',
            '**/*.config.{js,ts}',
          ]),
          optionalDependencies: false,
        },
      ],
      'import/order': [
        'error',
        {
          alphabetize: {order: 'asc'},
          'newlines-between': 'always',
          pathGroups: [
            {pattern: 'src/**/*', group: 'parent', position: 'before'},
            {pattern: 'assets/**/*', group: 'parent', position: 'before'},
          ],
          pathGroupsExcludedImportTypes: ['builtin'],
        },
      ],
      'no-void': ['error', {allowAsStatement: true}],
      'prettier/prettier': 'error',
      'sort-imports': ['error', {ignoreDeclarationSort: true}],
      ...parserRules(false, isReact),
    },
    overrides: [
      {
        files: ['**/*.ts?(x)'],
        extends: [
          'plugin:@typescript-eslint/recommended-requiring-type-checking',
        ],
        rules: {
          ...parserRules(true, isReact),
        },
      },
      {
        files: testMatch,
        rules: {
          'no-empty': ['error', {allowEmptyCatch: true}],
        },
      },
      {
        files: ['**/*/__tests__/helpers/**/*', '**/*/__tests__/utils/**/*'],
        rules: {
          'jest/no-export': 'off',
        },
      },
    ],
  }
}

module.exports = {buildConfig}
