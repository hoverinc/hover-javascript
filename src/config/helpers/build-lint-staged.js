const {resolveHoverScripts, resolveBin} = require('../../utils')

const hoverScripts = resolveHoverScripts()
const doctoc = resolveBin('doctoc')

const defaultTestCommand = `${hoverScripts} test --findRelatedTests --passWithNoTests`

const sourceExtensions = ['js', 'jsx', 'ts', 'tsx']

const readmeGlob = 'README.md'
const formatGlob =
  '*.+(js|jsx|json|json5|yml|yaml|css|less|scss|ts|tsx|md|graphql|mdx|vue)'

// This works around the limitation imposed by using globs as keys in the
// configuration. We want to run the lint and test commands on the same glob,
// but we want to do so in parallel. If we specify the commands in an array
// under the same glob key, they always run in sequence.
const lintGlob = `*.+(${sourceExtensions.join('|')})`
const testGlob = `*.+(${sourceExtensions.reverse().join('|')})`

const buildConfig = (toc = true, testCommand = defaultTestCommand) => ({
  ...(toc ? {[readmeGlob]: [`${doctoc} --maxlevel 4 --notitle`]} : {}),
  [formatGlob]: [`${hoverScripts} format`],
  [lintGlob]: [`${hoverScripts} lint`],
  [testGlob]: [testCommand],
})

module.exports = {buildConfig}
