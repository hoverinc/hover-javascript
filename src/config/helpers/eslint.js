const {rules} = require('eslint-config-airbnb-typescript/lib/shared')

const {hasAnyDep} = require('../../utils')
const {testMatch} = require('../jest.config')

const withBaseConfig = base => variant =>
  require.resolve(base + (variant ? `/${variant}` : ''))

const airbnb = withBaseConfig('eslint-config-airbnb-typescript')
const prettier = withBaseConfig('eslint-config-prettier')

const hasReact = hasAnyDep('react')

const parserRules = (typescript = false) => {
  const isOff = off => (off ? 'off' : 'error')

  return {
    'no-implied-eval': isOff(typescript),
    'no-throw-literal': isOff(typescript),
    '@typescript-eslint/no-implied-eval': isOff(!typescript),
    '@typescript-eslint/no-throw-literal': isOff(!typescript),
  }
}

const buildConfig = ({withReact = false} = {}) => {
  const ifReact = (t, f) => (withReact || hasReact ? t : f)

  return {
    plugins: ['prettier', 'jest', ifReact('react-hooks')].filter(Boolean),
    extends: [
      ifReact(airbnb(), airbnb('base')),
      'plugin:jest/recommended',
      prettier(),
      prettier('@typescript-eslint'),
      ifReact(prettier('react')),
    ].filter(Boolean),
    rules: {
      'prettier/prettier': 'error',
      'import/prefer-default-export': 'off',
      'import/no-extraneous-dependencies': [
        'error',
        {
          devDependencies: rules[
            'import/no-extraneous-dependencies'
          ][1].devDependencies.concat([
            'jest/**',
            'e2e/**',
            '**/prettier.config.js',
          ]),
          optionalDependencies: false,
        },
      ],
      ...parserRules(),
    },
    overrides: [
      {
        files: ['**/*.ts?(x)'],
        parserOptions: {
          project: './tsconfig.json',
        },
        extends: [
          'plugin:@typescript-eslint/recommended-requiring-type-checking',
        ],
        rules: {
          ...parserRules(true),
        },
      },
      {
        files: testMatch,
        rules: {
          'no-empty': ['error', {allowEmptyCatch: true}],
        },
      },
    ],
  }
}

module.exports = {buildConfig}
