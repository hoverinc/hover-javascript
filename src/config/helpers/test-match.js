const toGlob = extensions => `*.+(${extensions.join('|')})`

const testMatchExtensions = ['js', 'jsx', 'ts', 'tsx']
const testMatchGlob = toGlob(testMatchExtensions)
const testMatchSuffixGlob = toGlob(
  testMatchExtensions.map(extension => 'test.'.concat(extension)),
)

const testMatch = [
  `**/__tests__/**/${testMatchGlob}`,
  `test/**/${testMatchGlob}`,
  `test/**/${testMatchSuffixGlob}`,
  `e2e/**/${testMatchSuffixGlob}`,
  `**/${testMatchSuffixGlob}`,
]

module.exports = {
  testMatchExtensions,
  testMatchGlob,
  testMatchSuffixGlob,
  testMatch,
}
