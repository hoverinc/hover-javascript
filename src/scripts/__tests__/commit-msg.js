import cases from 'jest-in-case'
import {unquoteSerializer, winPathSerializer} from './helpers/serializers'

expect.addSnapshotSerializer(unquoteSerializer)
expect.addSnapshotSerializer(winPathSerializer)

cases(
  'commit-msg',
  ({
    args = [],
    utils = require('../../utils'),
    hasFile = () => false,
    env = {},
  }) => {
    // beforeEach
    const {sync: crossSpawnSyncMock} = require('cross-spawn')
    const originalArgv = process.argv
    const originalExit = process.exit
    const originalEnv = process.env

    Object.assign(utils, {
      hasFile,
      resolveBin: (modName, {executable = modName} = {}) => executable,
    })

    process.exit = jest.fn()
    process.argv = ['node', '../commit-msg', ...args]
    process.env = env

    try {
      // tests
      require('../commit-msg')

      expect(crossSpawnSyncMock).toHaveBeenCalledTimes(1)

      const [call] = crossSpawnSyncMock.mock.calls
      const [script, calledArgs] = call

      expect([script, ...calledArgs].join(' ')).toMatchSnapshot()
    } catch (error) {
      throw error
    } finally {
      // afterEach
      process.exit = originalExit
      process.argv = originalArgv
      process.env = originalEnv

      jest.resetModules()
    }
  },
  {
    'calls @commitlint/cli with default args': {},
    'does not use built-in config with --config': {
      args: ['--config', './custom-config.js'],
    },
    'does not use built-in config with commitlint.config.js file': {
      hasFile: filename => filename === 'commitlint.config.js',
    },
    'forwards args': {
      args: ['--verbose'],
    },
    'adds env flag with HUSKY_GIT_PARAMS when available': {
      env: {HUSKY_GIT_PARAMS: 'husky-git-params'},
    },
  },
)
