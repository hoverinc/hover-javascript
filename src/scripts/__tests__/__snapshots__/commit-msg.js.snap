// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commit-msg adds env flag with HUSKY_GIT_PARAMS when available 1`] = `commitlint --env HUSKY_GIT_PARAMS --config ./src/config/commitlint.config.js`;

exports[`commit-msg calls @commitlint/cli with default args 1`] = `commitlint --config ./src/config/commitlint.config.js`;

exports[`commit-msg does not use built-in config with --config 1`] = `commitlint --config ./custom-config.js`;

exports[`commit-msg does not use built-in config with commitlint.config.js file 1`] = `commitlint`;

exports[`commit-msg forwards args 1`] = `commitlint --config ./src/config/commitlint.config.js --edit .git/COMMIT_EDITMSG`;
